pubsublite-kafka-shim/src/main/java/com/google/cloud/pubsublite/kafka/SingleSubscriptionConsumerImpl.java

```diff
@@ -0,0 +1,288 @@
+
+  private final TopicPath topic;
+  private final boolean autocommit;
+
+  private final PullSubscriberFactory subscriberFactory;
+  private final CommitterFactory committerFactory;
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
+
+  static class SubscriberState {
+    PullSubscriber<SequencedMessage> subscriber;
+    Committer committer;
+    Optional<Offset> lastUncommitted = Optional.empty();
+  }
+
+  @GuardedBy("monitor.monitor")
+  private Map<Partition, SubscriberState> partitions = new HashMap<>();
+
+  // Set to true when wakeup() has been called once.
+  @GuardedBy("monitor.monitor")
+  private boolean wakeupTriggered = false;
+
+  SingleSubscriptionConsumerImpl(
+      TopicPath topic,
+      boolean autocommit,
+      PullSubscriberFactory subscriberFactory,
+      CommitterFactory committerFactory) {
+    this.topic = topic;
+    this.autocommit = autocommit;
+    this.subscriberFactory = subscriberFactory;
+    this.committerFactory = committerFactory;
+  }
+
+  @Override
+  public void setAssignment(Set<Partition> assignment) {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      List<SubscriberState> unassigned =
+          ImmutableSet.copyOf(partitions.keySet()).stream()
+              .filter(p -> !assignment.contains(p))
+              .map(p -> partitions.remove(p))
+              .collect(Collectors.toList());
+      for (SubscriberState state : unassigned) {
+        state.subscriber.close();
+        state.committer.stopAsync().awaitTerminated();
+      }
+      assignment.stream()
+          .filter(p -> !partitions.containsKey(p))
+          .forEach(
+              ExtractStatus.rethrowAsRuntime(
+                  partition -> {
+                    SubscriberState s = new SubscriberState();
+                    s.subscriber =
+                        subscriberFactory.newPullSubscriber(
+                            partition,
+                            SeekRequest.newBuilder()
+                                .setNamedTarget(NamedTarget.COMMITTED_CURSOR)
+                                .build());
+                    s.committer = committerFactory.newCommitter(partition);
+                    s.committer.startAsync().awaitRunning();
+                    partitions.put(partition, s);
+                  }));
+    } catch (Throwable t) {
+      throw ExtractStatus.toCanonical(t).getStatus().asRuntimeException();
+    }
+  }
+
+  @Override
+  public Set<Partition> assignment() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return partitions.keySet();
+    }
+  }
+
+  @Override
+  public ConsumerRecords<byte[], byte[]> poll(Duration duration) {
+    Map<Partition, Queue<SequencedMessage>> partitionQueues = new HashMap<>();
+    try {
+      if (wakeupTriggered) throw new WakeupException();
+      while (!duration.isZero()) {
+        Duration sleepFor = Collections.min(ImmutableList.of(duration, Duration.ofMillis(100)));
```

Comment:
I'm worried that this will result in very high e2e latency for consumers that set high poll durations. IIUC, Kafka returns from Poll as soon as any data is available, and does not wait until "max.poll.records" records are available. Ideally, poll() would return as soon as any one partition has any data unless "fetch.min.bytes" is set to >1.

Reply:
Yes, I was misreading the behavior. I've restructured this to return as soon as any messages are available, PTAL.

Reply:
This works well for cases where messages are already in the buffer when the client calls poll(), but it will still have high latency when messages are delivered during the 100ms sleep. I can think of some cases where this will matter: imagine messages being published uniformly every 10ms. Also, we need to remember that clients are used to single-digit milliseconds [latency](https://stackoverflow.com/a/35071082) for Kafka. Would it be possible to have some kind of notification when any of the SinglePartitionSubscribers has messages ready to be delivered?